### PR TITLE
Replace Hangfire with custom CRON job runner

### LINKED ISSUE #208: CRON service
Hangife seems too complicated for usage and stores a lot of data to Db, which is not necessary.
And we had much more issues with it.
I propose to create a simple CRON service, that will have:

- Jobs that can set repeat time
- One-time jobs (?)
- Persistent storage, aka registered Jobs implementations, will use a database and will start after program startup
  - On cron schedule event, Producer will add the job into a _job queue_ in persistent storage and will be processed by a Consumer
  - The Job record will include some states like:
    - `Todo`, `Processing`, `Failed`, `Successful -- not sure if we need the later 1 or 2 tho, we can report failed to log and remove successful ones.
    - Path to class, hangfire stores this: `{"Type":"HonzaBotner.Discord.Services.Jobs.TriggerRemindersJobProvider, HonzaBotner.Discord.Services, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null","Method":"Run","ParameterTypes":"[]","Arguments":"[]"}`
    - Created at

### PR DESCRIPTION
**Pull request by ostorc**
Closes #208 

Todo: 

- [ ] migrate Hangire jobs
- [ ] add some statistics for jobs (not sure yet)

### CODE REVIEW

**Comment by ostorc** [comment]
Example implementations:

Scoped job:
```csharp
[Cron("*/10 * * * * *")] // Not mandatory, you can register it with cron explicitly
public class WarningCountJob : IJob
{
    private readonly IWarningService _warningService;
    private readonly ILogger<WarningCountJob> _logger;
    public string Name { get; } = "Warning Count Job";

    public WarningCountJob(IWarningService warningService, ILogger<WarningCountJob> logger)
    {
        _warningService = warningService;
        _logger = logger;
    }

    public async Task ExecuteAsync(CancellationToken cancellationToken = default)
    {
        var warnings = await _warningService.GetAllWarningsAsync();

        _logger.LogError("Total warning count: {WarningCount}", warnings.Count);
    }
}
```

Singleton job:

```csharp
public class PingJob : ICronJob
{
    private readonly ILogger<PingJob> _logger;
    public string Name { get; } = "Ping";
    public string CronExpression { get; } = "*/20 * * * * *";

    public PingJob(ILogger<PingJob> logger)
    {
        _logger = logger;
    }

    public Task ExecuteAsync(CancellationToken cancellationToken = default)
    {
        _logger.LogWarning("Ping at {Now}", DateTime.Now);
        return Task.CompletedTask;
    }
}
```

**Comment by ostorc** [comment]
How to configure services:
```csharp
services.AddScheduler(5000)
    .AddCronJob<PingJob>()
    .AddScopedCronJob<WarningCountJob>("*/20 * * * * *") // we can explicitly use cron expression
    .AddScopedCronJob<WarningCountJob>(); // or use CronAttribute
```

File: src/HonzaBotner.Scheduler/Scheduler.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using HonzaBotner.Scheduler.Contract;
+
+namespace HonzaBotner.Scheduler
+{
+    public class SchedulerHostedService : BackgroundService
```
Comment (tenhobi) [inline src/HonzaBotner.Scheduler/Scheduler.cs]: Class name doesn't match the file name.

File: src/HonzaBotner.Scheduler/Scheduler.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        protected override async Task ExecuteAsync(CancellationToken cancellationToken)
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                DateTime currentTime = DateTime.UtcNow;
+
+               _logger.LogInformation("Scheduler running at: {Time}", currentTime.ToLocalTime());
+               await RunOnceAsync(currentTime, cancellationToken);
+
+               await Task.Delay(_delay, cancellationToken);
```
Comment (tenhobi) [inline src/HonzaBotner.Scheduler/Scheduler.cs]: Why this would not stop working after some time like we tried the `Task.Delay` before?

**Review by tenhobi — COMMENTED** [review]
We are putting contracts to the same package now?

image

// https://github.com/fit-ctu-discord/honza-botner/issues/175

File: src/HonzaBotner.Scheduler/Scheduler.cs (hunk 2 above)
Comment (ostorc) [reply src/HonzaBotner.Scheduler/Scheduler.cs]: This should be Ok. I dont remember exact reason why the previous usage of Task.Delay failed, but here, we have the main worker thread, which only does two thing, spawns new seperate threads and waits for the delay. No real work is done here.

File: src/HonzaBotner.Scheduler/Scheduler.cs (hunk 1 above)
Comment (ostorc) [reply src/HonzaBotner.Scheduler/Scheduler.cs]: Fixed

**Comment by ostorc** [comment]
> We are putting contracts to the same package now?
> 
> // #175

Mainly, because I didnt want to add another package only with three files... And since this is fairly lightweight library, I dont think it will be issue.

**Review by tenhobi — APPROVED** [review]
LGTM

**Pull request merged**